File: chrome/android/java/src/org/chromium/chrome/browser/download/DownloadMessageBridge.java
// Copyright 2022 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.download;

import static org.chromium.build.NullUtil.assumeNonNull;

import android.content.Context;

import org.jni_zero.CalledByNative;

import org.chromium.base.JniOnceCallback;
import org.chromium.build.annotations.NullMarked;
import org.chromium.chrome.R;
import org.chromium.chrome.browser.ui.messages.snackbar.Snackbar;
import org.chromium.chrome.browser.ui.messages.snackbar.SnackbarManager;
import org.chromium.chrome.browser.ui.messages.snackbar.SnackbarManagerProvider;
import org.chromium.ui.base.WindowAndroid;

@NullMarked
public class DownloadMessageBridge {
    @CalledByNative
    public static void showIncognitoDownloadMessage(JniOnceCallback<Boolean> callback) {
        DownloadMessageUiController messageUiController =
                DownloadManagerService.getDownloadManagerService()
                        .getMessageUiController(/* otrProfileId= */ null);
        messageUiController.showIncognitoDownloadMessage(callback);
    }

    @CalledByNative
    public static void showUnsupportedDownloadMessage(WindowAndroid window) {
        SnackbarManager snackbarManager = SnackbarManagerProvider.from(window);
        if (snackbarManager == null) return;

        Context context = window.getContext().get();
        assumeNonNull(context);
        Snackbar snackbar =
                Snackbar.make(
                        context.getString(R.string.download_file_type_not_supported),
                        null,
                        Snackbar.TYPE_NOTIFICATION,
                        Snackbar.UMA_AUTO_LOGIN);
        snackbar.setAction(context.getString(R.string.ok), null);
        snackbar.setDefaultLines(false);
        snackbarManager.showSnackbar(snackbar);
    }
}

File: chrome/browser/download/android/download_message_bridge.cc
// Copyright 2022 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/download/android/download_message_bridge.h"

#include "base/android/jni_android.h"
#include "base/android/jni_callback.h"
#include "content/public/browser/web_contents.h"
#include "ui/android/window_android.h"

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "chrome/android/chrome_jni_headers/DownloadMessageBridge_jni.h"

DownloadMessageBridge::DownloadMessageBridge() = default;
DownloadMessageBridge::~DownloadMessageBridge() = default;

void DownloadMessageBridge::ShowIncognitoDownloadMessage(
    DownloadMessageRequestCallback callback) {
  JNIEnv* env = base::android::AttachCurrentThread();
  CHECK(!callback.is_null());
  // Convert the C++ callback to a JNI callback using ToJniCallback.
  Java_DownloadMessageBridge_showIncognitoDownloadMessage(
      env, base::android::ToJniCallback(env, std::move(callback)));
}

void DownloadMessageBridge::ShowUnsupportedDownloadMessage(
    content::WebContents* web_contents) {
  DCHECK(web_contents);
  JNIEnv* env = base::android::AttachCurrentThread();
  ui::WindowAndroid* window_android = web_contents->GetTopLevelNativeWindow();

  if (!window_android) {
    return;
  }

  Java_DownloadMessageBridge_showUnsupportedDownloadMessage(
      env, window_android->GetJavaObject());
}

DEFINE_JNI(DownloadMessageBridge)
